File: 1642-water-bottles/water-bottles.cpp
class Solution {
public:

    // Brute Force solution By Simulation
    int solution1(int numBottles, int numExchange){
        int consumed = 0;

        while(numBottles >= numExchange){

            consumed += numExchange;

            // Remove the numExchange bottles from the numBottles
            numBottles -= numExchange;

            // Add one Full bottle
            numBottles +=1;
        }

        return consumed+numBottles;
    }

    /*
        Analysis:
        Time Compelxity : O(N)
        Space Complexity : O(1)
    */    

    int solution2(int numBottles, int numExchange){
        int consumed = numBottles;
        int emptyBottles = numBottles;

        while(emptyBottles >= numExchange){

            int extraFullBottles = emptyBottles / numExchange;
            int remain = emptyBottles % numExchange;

            emptyBottles = extraFullBottles + remain;

            consumed +=extraFullBottles;
        }
        return consumed;
    }

    /*
        Analysis:
        Time Complexity : O(logN)
        Space Complexity : O(1)
    */    

    int numWaterBottles(int numBottles, int numExchange) {
        int ans = 0;

        ans += numBottles;
        while(numBottles>=numExchange){
            ans += (numBottles / numExchange);
            numBottles = numBottles%numExchange + (numBottles / numExchange);
        }

        return ans;


        // return solution2(numBottles, numExchange);
    }
};

File: 1642-water-bottles/water-bottles.java
class Solution {

    // Solution using simulations
    private int solution1(int numBottles, int numExchange) {
        int consumedBottles = numBottles;
        int currentBottles  = numBottles;

        while (currentBottles >= numExchange) {
            int emptyBottles = currentBottles / numExchange;
            consumedBottles += emptyBottles;

            currentBottles = (currentBottles % numExchange) + emptyBottles;
        }

        return consumedBottles;
    }
    /*
        Analysis:
            Time Complexity  : O(logn)
            Space Complexity : O(1)
    */

    // Solution using Math
    private int solution2(int numBottles, int numExchange) {
        return numBottles + ((numBottles - 1) / (numExchange - 1));
    }
    /*
        Analysis:
            Time Complexity  : O(1)
            Space Complexity : O(1)
    */

    public int numWaterBottles(int numBottles, int numExchange) {
        return solution2(numBottles, numExchange);
    }
}